File: message-queue/golang/src/telemetry/prometheus.go
package telemetry

import (
	"fmt"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/client_golang/prometheus/push"
	"golang.org/x/exp/maps"
	"strconv"
	"time"
)

import (
	"github.com/jan-matejka/code-golf/message-queue/golang/src/core"
)

var (
	TestMetric = prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Name: "test",
		Help: "Test metric",
	},
		append(append([]string{"worker_id"}, core.RuntimeFieldNames()...), core.SampleDescFieldNames()...),
	)

	MessagesTotal = prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Name: "messages_total",
		Help: "Messages sent",
	},
		append(append([]string{"worker_id"}, core.RuntimeFieldNames()...), core.SampleDescFieldNames()...),
	)

	MessagesPerSecond = prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Name: "messages_per_second",
		Help: "Messages per second sent",
	},
		append(append([]string{"worker_id"}, core.RuntimeFieldNames()...), core.SampleDescFieldNames()...),
	)

	DurationSeconds = prometheus.NewGaugeVec(prometheus.GaugeOpts{
		Name: "duration_seconds",
		Help: "Work duration in seconds",
	},
		append(append([]string{"worker_id"}, core.RuntimeFieldNames()...), core.SampleDescFieldNames()...),
	)
)

func newPusher() *push.Pusher {
	registry := prometheus.NewRegistry()
	registry.MustRegister(TestMetric)
	registry.MustRegister(MessagesTotal)
	registry.MustRegister(MessagesPerSecond)
	registry.MustRegister(DurationSeconds)

	pusher := push.New("http://localhost:9091", "mq-producer").Gatherer(registry)
	return pusher
}

func mkLabels(r *core.Runtime, sample core.SampleDesc, worker *core.WorkerResult) prometheus.Labels {
	str_labels := r.Map()
	maps.Copy(str_labels, map[string]string{"worker_id": strconv.Itoa(worker.WorkerId)})
	maps.Copy(str_labels, sample.Map())
	var labels prometheus.Labels = str_labels
	return labels
}

type PrometheusMetrics struct {
	pusher  *push.Pusher
	runtime *core.Runtime
}

func NewPrometheusMetrics(r *core.Runtime) PrometheusMetrics {
	pm := PrometheusMetrics{newPusher(), r}
	return pm
}

func (pm PrometheusMetrics) Push(sample core.SampleDesc, rs *core.Results) {
	for _, r := range rs.Workers {
		MessagesTotal.With(mkLabels(pm.runtime, sample, r)).Set(float64(r.MessagesTotal))
		MessagesPerSecond.With(mkLabels(pm.runtime, sample, r)).Set(r.MessagesPerSecond)
		DurationSeconds.With(mkLabels(pm.runtime, sample, r)).Set(r.DurationSeconds)

		if err := pm.pusher.Add(); err != nil {
			panic(fmt.Sprintf("Prometheus push failed: %v", err.Error()))
		}
	}
}

func (pm PrometheusMetrics) TestPush() {
	fmt.Println("Testing prometheus push")
	result := core.NewWorkerResult(0, 1, time.Second)
	labels := mkLabels(pm.runtime, core.SampleDesc{8, "goroutines", "postgres"}, result)
	TestMetric.With(labels).Set(42)
	if err := pm.pusher.Add(); err != nil {
		panic(fmt.Sprintf("Prometheus push failed: %v", err.Error()))
	}
}

File: message-queue/cpp/src/telemetry/prometheus.cpp
#ifndef TELEMETRY_PROMETHEUS_CPP
#define TELEMETRY_PROMETHEUS_CPP

#include "./prometheus.hpp"

namespace telemetry::prometheus {

Gateway mk_gateway(const Config& c) {
  auto job_name = string("mq-producer");
  return Gateway(
    c.prometheus_gateway_host,
    to_string(c.prometheus_gateway_port),
    job_name
  );
}

Prometheus::Prometheus(const Config &c)
: registry(std::make_shared<Registry>())
, gateway(mk_gateway(c))
, test_metric(
    BuildGauge()
    .Name("test")
    .Help("Test Metric")
    .Register(*registry)
  )
, messages_total(
    BuildGauge()
    .Name("messages_total")
    .Help("Messages sent")
    .Register(*registry)
  )
, messages_per_second(
    BuildGauge()
    .Name("messages_per_second")
    .Help("Messages per second sent")
    .Register(*registry)
  )
, duration_seconds(
    BuildGauge()
    .Name("duration_seconds")
    .Help("Work duration in seconds")
    .Register(*registry)
  )
{
  gateway.RegisterCollectable(registry);
}

Labels mk_labels(
  const Runtime& rt
, const WorkerResult& wr
, const SampleDesc& sdesc
) {
  Labels labels = rt.Map();
  labels.merge(sdesc.Map());
  labels.insert(pair{"worker_id", to_string(wr.WorkerId)});
  return labels;
}

void Prometheus::Push(
  const Runtime& rt
, const SampleDesc& sdesc
, const Results& rs
) {
  for(const auto& wr : rs.Workers) {
    messages_total
      .Add(mk_labels(rt, wr, sdesc))
      .Set(wr.MessagesTotal);

    messages_per_second
      .Add(mk_labels(rt, wr, sdesc))
      .Set(wr.MessagesPerSecond);

    duration_seconds
      .Add(mk_labels(rt, wr, sdesc))
      .Set(wr.DurationSeconds);
  }

  Push();
}

void Prometheus::Push() {
  auto status = gateway.PushAdd();
  if (status != 200) {
    THROW("Prometheus push failed: {}", status);
  }
}

void Prometheus::PushTest(const Runtime& rt) {
  INFO("Testing push to prometheus");
  INFO(string(rt));

  auto labels = mk_labels(
    rt,
    WorkerResult(0, 42, chrono::seconds(1)),
    SampleDesc(4, "threading", "postgres")
  );
  test_metric.Add(labels).Increment();
  Push();
}

}
#endif
